### DISCUSSION TITLE
zh/models/rtdetr/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#14411] [Docs]
# zh/models/rtdetr/

探索百度的RT-DETR ，这是一款基于视觉转换器的实时物体检测器，具有高精度和适应性强的推理速度。了解更多信息，请访问Ultralytics 。

https://docs.ultralytics.com/zh/models/rtdetr/

**Comment by abcc235** [comment]
你好，请问我想用RT-DETR训练分类任务该怎么配置tdetr-l.yaml网络？

**Reply by glenn-jocher** [reply]
@abcc235 你好！

要使用RT-DETR训练分类任务，您需要对模型配置文件进行一些调整。RT-DETR主要用于物体检测，但您可以通过修改配置文件来适应分类任务。以下是一个简单的示例，展示如何配置`rtdetr-l.yaml`文件：

1. **修改模型配置文件**：
   - 打开`rtdetr-l.yaml`文件。
   - 修改模型的输出层，使其适应分类任务。

2. **训练模型**：
   - 使用Ultralytics的Python API或CLI进行训练。

### 示例代码

**Python**:
```python
from ultralytics import RTDETR

# 加载预训练的RT-DETR-l模型
model = RTDETR("rtdetr-l.pt")

# 显示模型信息（可选）
model.info()

# 在COCO8示例数据集上训练模型100个epoch
results = model.train(data="coco8.yaml", epochs=100, imgsz=640)
```

**CLI**:
```bash
# 使用预训练的RT-DETR-l模型在COCO8示例数据集上训练100个epoch
yolo train model=rtdetr-l.pt data=coco8.yaml epochs=100 imgsz=640
```

请确保您使用的是最新版本的Ultralytics库，以避免已修复的bug。如果问题仍然存在，请提供一个最小可复现的示例代码，这将有助于我们更快地诊断和解决问题。您可以参考[最小可复现示例](https://docs.ultralytics.com/help/minimum_reproducible_example)来创建示例。

希望这些信息对您有所帮助！如果有其他问题，请随时告知 😊

**Comment by guansheng201226** [comment]
Hi, great work on the object detection model! Does it support segmentation as well? Thanks!

**Reply by guansheng201226** [reply]
你好，RT-Detr支持分割吗？

**Reply by pderrenger** [reply]
@guansheng201226 你好，感谢你的关注！目前，RT-DETR 主要支持物体检测任务，并不支持分割功能。如果你需要分割功能，可以考虑使用YOLOv8-seg模型。更多详细信息，请访问我们的[文档](https://docs.ultralytics.com/models/rtdetr/)。谢谢！

**Comment by juran123** [comment]
Hello, may I ask if a version with fewer RT-DETR parameters will be provided? Additionally, RT-DETR supports both attitude estimation and object detection simultaneously?

**Reply by glenn-jocher** [reply]
@juran123 currently, Ultralytics only provides pre-trained RT-DETR models focusing on object detection tasks, and these models do not support simultaneous pose estimation. For a reduced-parameter model, you may need custom modifications of RT-DETR. Feel free to explore the [RT-DETR documentation](https://docs.ultralytics.com/models/rtdetr/) for existing models and configurations.

**Comment by wall633** [comment]
你好，请问使用RT-DETR模型，该怎么设置超参数才能达到论文里面的水准？

**Reply by glenn-jocher** [reply]
@wall633 要达到论文中的RT-DETR表现，请参考官方提供的超参数设置，如训练数据集、学习率、批量大小等，并使用COCO预训练权重（如`rtdetr-l.pt`或`rtdetr-x.pt`）进行训练。详细的训练流程和代码示例请查看文档：https://docs.ultralytics.com/models/rtdetr/.

**Comment by ZombieeY** [comment]
你好，请问我想用RT-DETR目标检测，训练完成后使用训练好的best.pt在测试集进行测试，为什么会显示全为0呢

**Reply by glenn-jocher** [reply]
@ZombieeY 您好！训练后验证全为0的情况通常与数据集配置或训练过程有关。请按以下步骤排查：

1. 确保测试集的`data.yaml`正确指向图像路径，且标签文件与图像一一对应
2. 检查验证集标注格式是否符合YOLO要求（类别索引从0开始）
3. 训练时可添加`val=True`参数实时监控验证指标，参考RT-DETR验证模式文档

如果问题持续，建议提供完整的训练配置和数据集结构说明以便进一步分析。您也可以查看YOLO验证模式的最佳实践获取更多调试建议。

**Reply by ZombieeY** [reply]
你好，是这样的，这是我的训练与测试指令
yolo detect train data=gen.yaml model=rtdetr-l.pt epochs=72 batch=4 imgsz=640 device=0 workers=2 amp=False name=gen
yolo task=detect mode=val data=gen.yaml model=runs/detect/gen/weights/best.pt
用训练完成的best.pt or  last.pt 结果全部显示为0。然而，采用rtdetr-l.pt进行测试，尽管输出的类别是错误的，并且得到的测试结果为0.0001左右，起码能够显示，不会得到：
rt-detr-l summary: 502 layers, 31,991,960 parameters, 0 gradients, 103.4 GFLOPs
val: Scanning C:\Users\Administrator\Desktop\ultralytics-main\rust\labels\val.cache... 176 images, 0 backgrounds, 0 corrupt: 100%|██████████| 176/176 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:07<00:00,  1.48it/s]
                   all        176        363          0          0          0          0
Speed: 0.6ms preprocess, 21.9ms inference, 0.0ms loss, 0.6ms postprocess per image

**Reply by ZombieeY** [reply]
还有，我已经检查了标注格式，并且val=True，训练时的验证结果都能显示出来。非常期待您的回复。

**Reply by glenn-jocher** [reply]
@ZombieeY 您好！根据您的描述，训练时验证正常但测试全零的情况可能有以下原因：

1. **模型权重未正确保存**：请检查训练日志中保存的best.pt文件大小是否正常（应约106MB），并确保验证时使用了正确的权重路径

2. **测试集配置问题**：确认gen.yaml中val部分的路径指向真实测试集，且测试集与训练集的数据分布一致

3. **AMP兼容性问题**：虽然您已设置amp=False，但建议尝试升级到最新版Ultralytics代码库，该问题可能与早期版本中的混合精度训练稳定性有关

建议参考RT-DETR验证模式文档中的故障排除部分，并附上完整的训练日志截图以便进一步分析。如果问题持续存在，可以考虑在简化数据集（如COCO8）上复现训练过程来隔离问题。

根据AGPL-3.0许可要求，请注意任何使用Ultralytics模型进行训练/推理的行为都需要开源整个项目或获得Ultralytics企业授权，详见企业授权申请页面。

**Comment by ghj44544** [comment]
您好，我想问一下re-detr适用于仪表检测吗，如果不能，rt-detr有什么限定的条件吗（比如需要大量数据之类的）

**Reply by ghj44544** [reply]
也就是对仪表盘进行识别检测

**Reply by glenn-jocher** [reply]
RT-DETR适用于仪表检测任务，特别是在需要高精度和实时性能的场景。该模型通过混合编码器处理多尺度特征和IoU感知查询选择，适合复杂场景的检测（详见RT-DETR技术文档）。主要考量点包括：
1. 数据需求：与所有检测模型类似，建议准备充足标注数据（至少数百张典型场景样本），小样本时可尝试迁移学习
2. 硬件要求：推荐使用NVIDIA GPU+TensorRT加速以获得最佳实时性能，CPU推理速度会显著下降
3. 模型调整：可通过修改decoder层数平衡精度/速度（见RT-DETR自适应推理说明）

建议先用预训练模型进行迁移学习测试效果，具体训练方法可参考RT-DETR使用示例。

**Comment by anyhopex** [comment]
Do you support RT-DETRv2 ？

**Reply by glenn-jocher** [reply]
@qianliyx Currently Ultralytics only officially supports the original RT-DETR models (rtdetr-l.pt and rtdetr-x.pt) through our Python API and CLI. While we acknowledge RTDETRv2's improvements (released by Baidu in July 2024) in our documentation, we haven't yet integrated official support for RTDETRv2 models in our library. You can still explore and compare performance metrics of RTDETRv2 as mentioned in our comparison docs, but direct API support for RTDETRv2 isn't available at this time.

**Comment by yosoroQ** [comment]
Hello, can RE-DETR utilize solutions from Ultralytics YOLO, such as the 'Objects Counting in Regions' feature?

**Reply by yosoroQ** [reply]
Sorry，I entered it wrongly. It's RT-DETR.

**Reply by glenn-jocher** [reply]
@yosoroQ Yes, RT-DETR can leverage many Ultralytics solutions including object counting in regions. Since RT-DETR integrates with the Ultralytics ecosystem, you can use it with features like region-based counting, tracking, and other computer vision solutions. The workflow would be similar to YOLO models - use RT-DETR for detection and then apply the counting logic to specific regions of interest. You can find examples of object counting implementations in our [solutions documentation](https://docs.ultralytics.com/solutions/).

**Comment by kywish** [comment]
RT-DETRv3 has been released. Does Ultralytics have plans to integrate it?

**Comment by AK993** [comment]
你好，rt-detr支持在转为rknn格式，并在rk3588上部署吗

**Comment by liyansong123456** [comment]
RT-DETRv3 has been released. Does Ultralytics have plans to integrate it?

**Comment by yuwencool** [comment]
您好RT-DETR模型的推理支持model.track模式吗？

**Comment by wang18382140151-hub** [comment]
你好，最近deim,dfine等对于rtdetr有了很好的改进，尤其是deim能很好的加速训练的速度，能将其整合进rtdetr中吗？

**Comment by fxc0627** [comment]
你好，请问RE-DETR模型适用于小目标工业缺陷检测吗？对于小目标检测有什么天然的优势？

**Comment by maitianstudio-arch** [comment]
How to use RT-DETRv2?

**Reply by glenn-jocher** [reply]
Ultralytics currently exposes the original RT-DETR pretrained weights (`rtdetr-l.pt`, `rtdetr-x.pt`) via the `RTDETR` API; RT-DETRv2 pretrained weights/configs aren’t packaged in `ultralytics` yet, so if you share what RT-DETRv2 checkpoint you have (Paddle vs PyTorch) and how you’re trying to run it, we can point you to the right conversion/workflow—meanwhile you can use RT-DETR as shown in the [RT-DETR docs](https://docs.ultralytics.com/models/rtdetr/).

```python
from ultralytics import RTDETR

model = RTDETR("rtdetr-l.pt")
results = model("path/to/image.jpg")
```

**Comment by 995742853** [comment]
您好，我想请问一下，在哪里有基于rtdetr-resnet50的预训练pt权重，我知道看到rtdetr-l以及x的预训练权重

**Reply by glenn-jocher** [reply]
目前 [RT-DETR 文档](https://docs.ultralytics.com/zh/models/rtdetr/) 里公开提供的 Ultralytics 预训练权重只有 `rtdetr-l.pt` 和 `rtdetr-x.pt`；如果你找的是 ResNet50 版本，建议查看 Baidu 官方的 [RT-DETR PyTorch 权重列表](https://github.com/lyuwenyu/RT-DETR/tree/main/rtdetr_pytorch)，其中提供的是 `rtdetr_r50vd_*.pth`，不是 Ultralytics 直接发布的 `rtdetr-resnet50.pt`，也欢迎直接在这里继续提问，方便 Ultralytics 团队跟进。 ([docs.ultralytics.com](https://docs.ultralytics.com/zh/models/rtdetr/))

**Reply by 995742853** [reply]
谢谢您的回复

**Reply by glenn-jocher** [reply]
不客气，有其他 `RT-DETR` 使用问题欢迎直接继续在这里留言，方便 Ultralytics 团队一起查看。
